tools/generate_type_hints.py

```diff
@@ -0,0 +1,221 @@
+    "ubyte": "_UByteCodes",
+    "ushort": "_UShortCodes",
+    "uintc": "_UIntCCodes",
+    "uintp": "_UIntPCodes",
+    "uint": "_UIntCodes",
+    "ulonglong": "_ULongLongCodes",
+    "half": "_HalfCodes",
+    "single": "_SingleCodes",
+    "float_": "_DoubleCodes",
+    "longfloat": "_LongFloatCodes",
+    "csingle": "_CSingleCodes",
+    "complex_": "_CDoubleCodes",
+    "clongfloat": "_CLongFloatCodes",
+}
+
+TEMPLATE = r"""{docstring}
+
+from typing import Union, Any
+
+from numpy import (
+    int8,
+    int16,
+    int32,
+    int64,
+    uint8,
+    uint16,
+    uint32,
+    uint64,
+    float16,
+    float32,
+    float64,
+    complex64,
+    complex128,
+)
+
+from ._char_codes import (
+    _Int8CodesBase,
+    _Int16CodesBase,
+    _Int32CodesBase,
+    _Int64CodesBase,
+    _UInt8CodesBase,
+    _UInt16CodesBase,
+    _UInt32CodesBase,
+    _UInt64CodesBase,
+    _Float16CodesBase,
+    _Float32CodesBase,
+    _Float64CodesBase,
+    _Complex64CodesBase,
+    _Complex128CodesBase,
+    _ByteCodes,
+    _ShortCodes,
+    _IntCCodes,
+    _IntPCodes,
+    _IntCodes,
+    _LongLongCodes,
+    _UByteCodes,
+    _UShortCodes,
+    _UIntCCodes,
+    _UIntPCodes,
+    _UIntCodes,
+    _ULongLongCodes,
+    _HalfCodes,
+    _SingleCodes,
+    _DoubleCodes,
+    _LongFloatCodes,
+    _CSingleCodes,
+    _CDoubleCodes,
+    _CLongFloatCodes,
+)
+
+_Int8Codes = Union[_Int8CodesBase{int8}]
+_Int16Codes = Union[_Int16CodesBase{int16}]
+_Int32Codes = Union[_Int32CodesBase{int32}]
+_Int64Codes = Union[_Int64CodesBase{int64}]
+
+_UInt8Codes = Union[_UInt8CodesBase{uint8}]
+_UInt16Codes = Union[_UInt16CodesBase{uint16}]
+_UInt32Codes = Union[_UInt32CodesBase{uint32}]
+_UInt64Codes = Union[_UInt64CodesBase{uint64}]
+
+_Float16Codes = Union[_Float16CodesBase{float16}]
+_Float32Codes = Union[_Float32CodesBase{float32}]
+_Float64Codes = Union[_Float64CodesBase{float64}]
+
+_Complex64Codes = Union[_Complex64CodesBase{complex64}]
+_Complex128Codes = Union[_Complex128CodesBase{complex128}]
+
+# Note that these variables are private despite the lack of underscore;
+# don't expose them to the `numpy.typing` __init__ file.
+byte = {byte}
+short = {short}
+intc = {intc}
+intp = int0 = {intp}
+int_ = {int_}
+longlong = {longlong}
+
+ubyte = {ubyte}
+ushort = {ushort}
+uintc = {uintc}
+uintp = uint0 = {uintp}
+uint = {uint}
+ulonglong = {ulonglong}
+
+half = {half}
+single = {single}
+float_ = double = {float_}
+longfloat = longdouble = {longfloat}
+
+csingle = singlecomplex = {csingle}
+complex_ = cdouble = cfloat = {complex_}
+clongfloat = longcomplex = clongdouble = {clongfloat}
+"""
+
+
+def _file_to_context(
+    file: Union[IO[str], _AnyPath], *args: Any, **kwargs: Any
+) -> ContextManager[IO[str]]:
+    """Create a context manager from a path- or file-like object."""
+    try:
+        return open(file, *args, **kwargs)
+    except TypeError:
+        return nullcontext(file)
+
+
+def generate_alias(file: Union[IO[str], _AnyPath] = sys.stdout) -> None:
```

Comment:
I'd be inclined to remove this and either:

* Just use `open` directly in `generate_aliases`, since the only two places you call it pass strings
* Use `open` in the two callers, to match python APIs like `pickle.dump` and `json.dump`, both of which expect already open file handles.

Reply:
Sure, why not. I went with the second option in https://github.com/numpy/numpy/pull/17514/commits/368fc435bd68c1d098201bf3947efbd25a67babc.